blogposts/2021/postgres-version-update.md

```diff
@@ -0,0 +1,141 @@
+
+*lsif\_data\_definitions*:
+
+| index_id | scheme | identifier | data | version | 
+| -------- | ------ | ---------- | ---- | ------- | 
+| 1        | foo    | bar        | .... | 1       | 
+| 1        | bar    | baz        | .... | 2       | 
+| 2        | baz    | quux       | .... | 2       | 
+| 2        | quux   | bonk       | .... | 2       |
+| 3        | bonk   | honk       | .... | 1       |
+
+*lsif\_data\_definitions\_schema\_versions*:
+
+| index_id | min_version | max_version |
+| -------- | ----------- | ----------- |
+| 1        | 1           | 2           |
+| 2        | 2           | 2           |
+| 3        | 1           | 1           |
+
+This second table allows us to efficiently determine which indexes have been migrated and which still need attention. For example, if we are currently migrating from schema version 1 to 2, then the table above shows index 1 is partially migrated, index 2 is fully migrated, and index 3 has yet to be touched.
+
+It is an efficient operation to count both the total number of indexes as well as the number of indexes that are above or below a particular schema version.
+
+The schema version bounds can be kept in sync with the source table trivially by using triggers on inserts, updates, and deletes: if we insert a new row with a new lowest or new highest version or if we update/delete a row with the last remaining lowest or highest version, then we update the associated min or max version for that index.
+
+### The catch
+
+Of course this can't be the end of the story. In software engineering, every decision has a catch.
+
+After processing an LSIF index file, we insert data in Postgres utilizing very large bulk insert operations. During this process, we insert tens to hundreds of thousands of rows over multiple tables for each index - but we don't make tens of thousands of requests.
+
+For each insertion query, we supply as many rows as possible - up to `65535 / (# cols)` rows at a time. This reduces the total number of round trips to the database.
+
+As of this post, our [Cloud](https://sourcegraph.com/search) environment has a single index with ~390k rows in the definitions table alone, and each index has over 12k rows in this table on average. 12k rows can be inserted with a single statement, and 390k rows can be inserted over 30 statements.
+
+If we were to utilize a row-level trigger approach to keep the schema version bounds in sync with insertions, we would double our writes to the database: one insert in the target table and one upsert to the schema versions table.
+
+If we were instead to utilize a statement-level trigger approach, the extra cost would be negligible: only a handful of additional upsert operations. However, for this to work, we'd need to know the values of both the `index_id` and `version` column, which we do not have access to in Postgres 9.6.
```

Comment:
Add a conclusion summarizing the change and the reason for the change. This will do two things (1) remind the reader who got to the end of the article that we need them to take an action (2) ensures the post doesn't end abruptly.

Reply:
Added a quick thing here. Probably needs a better title.

Would like to point to docs on _how_ to perform the upgrade, but we haven't written those yet!

Reply:
@efritz that would be helpful. Is there a timeline for having that documentation?

Reply:
I'll know later today (I have a sync meeting on what needs to be done here).

Reply:
I should have links before the publish date to docs on sourcegraph.com.

Reply:
Updated.